### PR TITLE
Implemented CoverageExtractor API and register decorators (fixes #287)

### LINKED ISSUE #287: Provide support for different coverage tools
Currently, BugZoo only provides the ability to compute coverage for C and C++ programs, and uses `gcov` to do so. Going forward, BugZoo needs to provide support for different languages (e.g., Java and Python), as well as alternative coverage tools for a particular language.

To add support for different coverage tools, we should introduce a new `CoverageExtractor` class, replacing the existing `CoverageManager`, which computes coverage using a particular tool (where each tool is implemented as a `CoverageExtractor` subclass). `CoverageExtractor` instances should be able to accept arguments to allow them to be customised to particular programs. For example, the `GcovExtractor` constructor may accept a list of files that should be instrumented.

## Internal API

Based on my experience of working with `gcov` and `coverage.py`, I propose that each `CoverageExtractor` implements the following methods:

* `prepare(container: Container) -> None`: instruments the source code, if necessary, and rebuilds the binary with the appropriate instrumentation, if needed.
* `cleanup(container: Container) -> None`: removes instrumentation from the source code, and rebuilds the binary without instrumentation.
* `extract(container: Container, from_files: Optional[List[str]]) -> FileLineSet`: reads any coverage files from disk (inside the container) before destroying those coverage files and returning the set of lines that were covered. If `from_files` is provided, coverage information will only be returned for the specified files. **(Is this ability really all that useful? This can be done in a single method call on the `FileLineSet` object.)**
* `coverage(container: Container, tests: Iterable[TestCase], *, instrument: bool = True) -> TestSuiteCoverage`: uses a given container to computes coverage information for a collection of tests. If `instrument` is set to `True`, the container will be instrumented and deinstrumented at the beginning and end of the call, respectively. Alternatively, if `instrument` is set to False, the container will be neither instrumented nor deinstrumented; it is assumed that the user has already instrumented the container.

## External API

All coverage-related methods exposed by the `ContainerManager` would remain the same. Users of BugZoo's Python library shouldn't need to change their code (or, if they're using the `CoverageManager` directly, the changes necessary should be minimal).

## Manifest File Format

In addition to replacing `CoverageManager` with `CoverageExtractor`, we will also need to modify the manifest file format to allow users to specify which coverage tool should be used along with any arguments that should be provided to that tool.

Below are some examples of what the revised `coverage` section might look like:

```
coverage:
  type: gcov
  files-to-instrument:
    - foo.c
```

```
coverage:
   type: coverage.py
```

```
coverage:
  type: cobertura
```

We need to decide what to do if a user doesn't provide a `coverage` section, or if there is no `type` property within the `coverage` section. One approach could be to associate a default coverage extractor type with each language (e.g., `gcov` for C and `coverage.py` for Python). If a language doesn't have a default coverage extractor, a warning could be raised to remind the user that coverage can't be computed for the given language; we could also throw a `CoverageNotSupportedForLanguage` exception if the user attempts to call any of the coverage-related methods exposed by the container manager.

## Runtime Dependencies

Currently, BugZoo assumes that `gcovr` is installed inside the container. Ideally, it shouldn't be necessary for the user to install their coverage tool of choice inside the container. Each `CoverageExtractor` could be optionally associated with a single `Plugin`, which is used to mount a set of portable libraries and binaries inside the container at launch. For example, `GCovExtractor` may be associated with a `gcovr` plugin which imports a portable, single-file `gcovr` binary into the container at `/opt/gcovr/bin/gcovr`.

## Roadmap

* [x] refactor core data structures to use `attrs` (#210)
* [x] add coverage instructions data structure
* [x] add hardcoded `CoverageExtractor` for `gcov`
* [x] remove server-side `CoverageManager` and fold logic into server-side `ContainerManager`
* [ ] rename `instrument` and `deinstrument` to `prepare` and `cleanup`.

### PR DESCRIPTION
**Pull request by ChrisTimperley**
**TODO:**

* [x] update coverage documentation
* [x] update manifest file format documentation
* [x] update "adding a bug" documentation
* [x] update exceptions documentation

### CODE REVIEW

File: bugzoo/core/bug.py [hunk 1]
```diff
@@ -42,16 +47,30 @@ def harness(self) -> TestSuite:
...
+                         name, d['coverage']['type'])
+            instructions_coverage = \
+                CoverageInstructions.from_dict(d['coverage'])
+        elif len(languages) == 1:
+            lang = languages[0]
+            logger.debug("bug [%s]: using default coverage instructions type for language [%s]",
+                         name, lang.name)
+            cls_cov_ins = CoverageInstructions.language_default(lang)
+            instructions_coverage = \
+                cls_cov_ins.from_dict(d.get('coverage', {}))
```
Comment (afsafzal) [inline bugzoo/core/bug.py]: What if `cls_cov_ins` is `None`?

File: bugzoo/core/coverage.py [hunk 2]
```diff
@@ -1,34 +1,108 @@
...
+
+    @classmethod
+    def register(cls, name: str) -> None:
+        logger.debug("registering coverage instructions [%s] under name [%s]",
+                     cls, name)
+        if name in _NAME_TO_INSTRUCTIONS:
+            raise exceptions.NameInUseError(name)
+        if cls in _INSTRUCTIONS_TO_NAME:
+            m = "coverage instructions already registered under name: {}"
+            m = m.format(_INSTRUCTIONS_TO_NAME[cls])
```
Comment (afsafzal) [inline bugzoo/core/coverage.py:64]: Forgot to log `m`, or raise exception

File: bugzoo/mgr/coverage/extractor.py [hunk 3]
```diff
@@ -9,265 +9,96 @@
...
+              container: Container
+              ) -> 'CoverageExtractor':
+        bug = installation.bugs[container.bug]  # type: Bug
+        instructions = bug.instructions
+        if instructions is None:
+            raise exceptions.NoCoverageInstructions
+
+        name = instructions.__class__.registered_under_name()
+        extractor_cls = _NAME_TO_EXTRACTOR[name]
+        extractor = extractor_cls(installation, container, instructions)
```
Comment (afsafzal) [inline bugzoo/mgr/coverage/extractor.py]: You expect `CoverageExtractor` to take these 3 arguments in `__init__` but that's not specified anywhere. If I want to write a `CoverageExtractor` I need to look at an example to understand what's happening.

File: bugzoo/mgr/coverage/extractor.py [hunk 4]
```diff
@@ -9,265 +9,96 @@
...
-                                   verbose=True)
-        logger_c.debug("Finished running gcovr (took %.2f seconds).", timer() - t_start)  # noqa: pycodestyle
-        assert response.code == 0, "failed to run gcovr"
+class CoverageExtractor(object):
+    """
+    Used to compute coverage information for a given container according to a
+    provided set of instructions.
+    """
+    @classmethod
+    def register(cls, name: str) -> None:
```
Comment (afsafzal) [inline bugzoo/mgr/coverage/extractor.py:100]: This function and the next one look very similar to the ones in `CoverageInstruction`. It's a bit confusing what an extractor is and what an instruction is and who is registering the name.

File: bugzoo/mgr/coverage/gcov.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        cmd = ' -o '.join(["-name \*{}".format(e) for e in endings])
+        cmd = "find {} -type f \( {} \)".format(dir_source, cmd)
+        resp = mgr_ctr.command(container, cmd)
+        all_files = set(fn.strip() for fn in resp.output.split('\n'))
+
+        def has_file(fn_rel: str) -> bool:
+            fn_abs = os.path.join(dir_source, fn_rel)
+            return (fn_abs in all_files)
+
+        def read_line_coverage(cls) -> List[int]:
```
Comment (afsafzal) [inline bugzoo/mgr/coverage/gcov.py]: `cls` type?

File: bugzoo/mgr/coverage/gcov.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        coverage = TestSuiteCoverage(cov)
+        logger.debug("Computed coverage for container: %s", container.uid)
+        return coverage
+
+    def cleanup(self) -> None:
+        """
+        Strips instrumentation from the source code inside the container,
+        and reconfigures its program to no longer use coverage options.
+        """
+        logger.debug("cleanup method for gcov extractor does nothing.")
```
Comment (afsafzal) [inline bugzoo/mgr/coverage/gcov.py:217]: Shouldn't it deinstrument?

File: bugzoo/mgr/coverage/extractor.py (hunk 4 above)
Comment (afsafzal) [reply bugzoo/mgr/coverage/extractor.py:100]: Also after reading `GCovExtractor` I notice that there are function required by the API to be implemented by `CoverageExtractor`'s subclasses. Those should be part of `CoverageExtractor`.

File: bugzoo/mgr/source.py [hunk 7]
```diff
@@ -168,17 +168,11 @@ def __parse_blueprint(self, source: Source, fn: str, d: dict) -> BuildInstructio
                                  source.name)
 
     def __parse_bug(self, source: Source, fn: str, d: dict) -> Bug:
```
Comment (afsafzal) [inline bugzoo/mgr/source.py:170]: is `fn` used anywhere?

File: bugzoo/mgr/coverage/extractor.py [hunk 8]
```diff
@@ -9,265 +9,96 @@
...
-        res = self._parse_report(report)
-        logger_c.debug("Finished parsing gcovr XML report (took %.2f seconds).", timer() - t_start)  # noqa: pycodestyle
-        logger_c.debug("Finished extracting coverage information")
-        return res
+    @staticmethod
+    def build(installation: 'BugZoo',
+              container: Container
+              ) -> 'CoverageExtractor':
+        bug = installation.bugs[container.bug]  # type: Bug
+        instructions = bug.instructions
```
Comment (afsafzal) [inline bugzoo/mgr/coverage/extractor.py]: I think this should be `bug. instructions_coverage`

**Review by afsafzal — CHANGES_REQUESTED** [review]

File: bugzoo/core/coverage.py (hunk 2 above)
Comment (ChrisTimperley) [reply bugzoo/core/coverage.py:64]: :+1:  added below

File: bugzoo/core/bug.py (hunk 1 above)
Comment (ChrisTimperley) [reply bugzoo/core/bug.py]: Nice spot. Added logic to handle this case.

File: bugzoo/mgr/source.py (hunk 7 above)
Comment (ChrisTimperley) [reply bugzoo/mgr/source.py:170]: Yes and no. In general, we use `fn` to figure out the directory that contains the bug manifest. We then use that directory to determine the absolute path to the `Dockerfile` (and build context) specified in the manifest. However, we only need that logic for blueprints, and not necessarily for bugs.  For the sake of consistency, I think it's worth keeping `fn` in the `__parse_bug` header for now, even if it isn't used directly.

File: bugzoo/mgr/coverage/gcov.py (hunk 6 above)
Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/gcov.py:217]: It should deinstrument, but the code that we had for doing that didn't really work and was disabled accordingly. I'll open a separate issue for this.

Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/gcov.py:217]: #302

File: bugzoo/mgr/coverage/extractor.py (hunk 4 above)
Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/extractor.py:100]: > Also after reading `GCovExtractor` I notice that there are function required by the API to be implemented by `CoverageExtractor`'s subclasses. Those should be part of `CoverageExtractor`.

I'm not quite sure what you mean. Could you elaborate?

File: bugzoo/mgr/coverage/gcov.py (hunk 5 above)
Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/gcov.py]: Renamed to `cobertura_class`.

File: bugzoo/mgr/coverage/extractor.py (hunk 3 above)
Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/extractor.py]: Added instructions to `CoverageExtractor` docstring.

File: bugzoo/mgr/coverage/extractor.py (hunk 4 above)
Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/extractor.py:100]: > This function and the next one look very similar to the ones in `CoverageInstruction`. It's a bit confusing what an extractor is and what an instruction is and who is registering the name.

We could rename `register` and `register_as_default` to `_register` and `_register_as_default`? These aren't supposed to be called by the user. Instead, the user should call the `@register` and `@register_as_default` decorators.

As for why this is necessary, we need a way to map `type` names in the manifest file to `CoverageInstruction` subclasses, which we then in turn need to map to their respective `CoverageExtractor` class. The implementation details are hidden from the user by requiring that they write a `CoverageExtractor` (with an internal `Instructions` class) and decorate that `CoverageExtractor` only once.

Comment (ChrisTimperley) [reply bugzoo/mgr/coverage/extractor.py:100]: > > Also after reading `GCovExtractor` I notice that there are function required by the API to be implemented by `CoverageExtractor`'s subclasses. Those should be part of `CoverageExtractor`.
> 
> I'm not quite sure what you mean. Could you elaborate?

I know what you mean. I've added abstract methods to `CoverageExtractor` and documented the API.

**Review by afsafzal — APPROVED** [review]

**Pull request merged**